### PR TITLE
lightning: catch auto recoverable error

### LINKED ISSUE #33624: Refinement of lightning error
## Enhancement
Lightning tries automatically recovering from some retryable errors. Two function called `IsRetryableError` and `MessageIsRetryableStorageError` could provide us a hint to distinguish recoverable and unrecoverable errors. However, though they could check whether an error is retryable or not, in some situation, retryable error should also be caught rather than ignored (i.e. `recoverable != retryable`)

It's necessary to refine the rules to check if an error is recoverable.

### PR DESCRIPTION
**Pull request by buchuitoudegou**
### What problem does this PR solve?

Issue Number: close #33624 

Problem Summary:

### What is changed and how it works?
Add function `IsAutoRecoverableError` to catch auto recoverable error. Currently, these errors are caught by `IsRetryableError` which sometimes produces false positive (i.e. some retryable error cannot be recovered). 

`IsAutoRecoverableError` catches network issues (which are recoverable in most situation).

### Check List

Tests 

- [x] Unit test

Side effects

Documentation

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by buchuitoudegou** [comment]
will add ut later.

File: br/pkg/utils/retry.go [hunk 1]
```diff
@@ -104,6 +104,33 @@ func IsRetryableError(err error) bool {
 	return true
 }
 
+// IsRecoverableError attempts to catch network issue especially timeout error
+// in order to resume lightning automatically
```
Comment (gozssky) [inline br/pkg/utils/retry.go:108]: If this method is used by Lightning only, I think we can move this method to Lightning's common package.

File: br/pkg/utils/retry.go [hunk 2]
```diff
@@ -104,6 +104,33 @@ func IsRetryableError(err error) bool {
...
+func IsAutoRecoverableError(err error) bool {
+	for _, singleError := range errors.Errors(err) {
+		if !isSingleAutoRecoverableError(singleError) {
+			return false
+		}
+	}
+	return true
+}
+
+func isSingleAutoRecoverableError(err error) bool {
```
Comment (gozssky) [inline br/pkg/utils/retry.go:118]: It seems we don't check all recoverable errors here.

**Comment by D3Hunter** [comment]
maybe enchance this directly
https://github.com/pingcap/tidb/blob/b9635e2792c557dc33376dbd40ca78a51d7a29fc/br/pkg/utils/retry.go#L132-L136

all error in this branch `case codes.Unknown:` should not be recoverable except some errors we can be sure it is recoverable

**Comment by buchuitoudegou** [comment]
> all error in this branch `case codes.Unknown:` should not be recoverable except some errors we can be sure it is recoverable, such as `mysql.ErrInvalidConn` `driver.ErrBadConn`

Did you mean we should do that like this:
```go
func IsAutoRecoverableError(err error) bool {
    if isSingleRetryableError(err) && (!not in the branch `code.Unknown` || isMySQLErrInvalidConn || isDriverErrBadConn) {
        return true
    }
    return false
}
```
Looks a bit dizzy😵‍💫

**Comment by D3Hunter** [comment]
> > all error in this branch `case codes.Unknown:` should not be recoverable except some errors we can be sure it is recoverable, such as `mysql.ErrInvalidConn` `driver.ErrBadConn`
> 
> Did you mean we should do that like this:
> 
> ```go
> func IsAutoRecoverableError(err error) bool {
>     if isSingleRetryableError(err) && (!in_the_branch_`code.Unknown` || isMySQLErrInvalidConn || isDriverErrBadConn) {
>         return true
>     }
>     return false
> }
> ```
> 
> Looks a bit dizzy😵‍💫

like in https://github.com/pingcap/tidb/pull/33510, but with this bug https://github.com/pingcap/tidb/pull/33616 fixed

**Comment by buchuitoudegou** [comment]
> like in #33510, but with this bug #33616 fixed

In #33616, `IsRetryableError` was simply removed. Did you mean I should reuse and refine the condition? For instance:

```go
switch err {
case IsRetryableError(err) && (!in_the_branch_`code.Unknown` || isMySQLErrInvalidConn || isDriverErrBadConn):
    return nil
}
```
As the original issue #33363 only would like to resume Lightning automatically from checkpoint when it crashed due to network error, using `IsRetryableError` will ignore more errors beside network timeout (e.g. `tmysql.ErrInfoSchemaChanged`). I don't think Lightning is supposed to resume itself from checkpoint if encountering this last time.

To be more succinct, retryable error doesn't equal auto-recoverable error. It's necessary and intuitive to write a new function to check it.  cc @gozssky

**Comment by D3Hunter** [comment]
what's `ErrInfoSchemaChanged`? seems we should not retry on `ErrInfoSchemaChanged` too

> retryable error doesn't equal auto-recoverable error.

what's the difference?

**Comment by D3Hunter** [comment]
maybe no need to change `IsRetryableError` in this pr, also need to modify it in https://github.com/pingcap/tidb/pull/33794/files

**Comment by buchuitoudegou** [comment]
> what's `ErrInfoSchemaChanged`? seems we should not retry on `ErrInfoSchemaChanged` too
> 
> 
> 
> > retryable error doesn't equal auto-recoverable error.
> 
> 
> 
> what's the difference?

I'm not sure where it's raised. But it will be caught by the IsRetryableError. So it would be life-saving to make another function rather than use IsRetryableError.

Recoverable error means the error is transient, happens occasionally, and doesn't result from "data error", while retryable error may not be auto-recoverable because, for instance, SchemaNotMatched might result from the schema is permanently changed or temporarily changed, where the former is recoverable and possibly attributes to some network lag but the latter is not (due to the defects of the system). If it is the latter one, we shouldn't resume it.

**Pull request closed without merging**